### PR TITLE
Dictionary-related optimizations

### PR DESCRIPTION
**Pull request by eric-wieser**
xref #961

One performance improvement on all versions of python, and an extra one for just the latest versions.

### CODE REVIEW

File: cocotb/scheduler.py [hunk 1]
```diff
@@ -71,6 +72,14 @@
...
 from cocotb.utils import nullcontext
 
+# On python 3.7 onwards, `dict` is guaranteed to preserve insertion order.
+# Since `OrderedDict` is a little slower that `dict`, we prefer the latter
+# when possible.
+if sys.version_info[:2] >= (3, 7):
+    _ordered_dict = dict
+else:
+    _ordered_dict = collections.OrderedDict
+
```
Comment (imphil) [inline cocotb/scheduler.py]: It's surprising to me that Python wouldn't do that optimization by itself.

**Comment by andresdemski** [comment]
```
 url https://github.com/potentialventures/cocotb/
 commit: master
 path: /tmp/cocotb-master
 results: 
    simtime: 270162.40 ns
    realtime: 12.40s
    performance: 21780.49 ns/s


 url https://github.com/eric-wieser/cocotb/
 commit: dict-optimizations
 path: /tmp/cocotb-eric-wieser
 results: 
    simtime: 270162.40 ns
    realtime: 11.71s
    performance: 23069.63 ns/s
```

performance improvement: ~6% :clap:

**Comment by eric-wieser** [comment]
@andresdemski: which python version?

File: cocotb/scheduler.py (hunk 1 above)
Comment (eric-wieser) [reply cocotb/scheduler.py]: OrderedDict has extra methods (that we don't use anyway), so I don't think it can

**Comment by andresdemski** [comment]
Those results are using python 3.7.

# python2.7 (in docker)
```
 url https://github.com/potentialventures/cocotb/
 commit: master
 path: /tmp/cocotb-master
 results: 
    simtime: 270162.40 ns
    realtime: 24.21s
    performance: 11157.73 ns/s


 url https://github.com/eric-wieser/cocotb/
 commit: dict-optimizations
 path: /tmp/cocotb-eric-wieser
 results: 
    simtime: 270162.40 ns
    realtime: 23.56s
    performance: 11466.81 ns/s
```
performance improvement: ~3%

# python 3.5 (in docker)
```
./run_in_docker.sh python3 compare.py                                                                                                                                                                       1 ↵
sha256:ad2ec252077490f5479f9fe611473e7ad90befbd67a89e49b68d54fa4c5a89e3

 url https://github.com/potentialventures/cocotb/
 commit: master
 path: /tmp/cocotb-master
 results: 
    simtime: 270162.40 ns
    realtime: 19.41s
    performance: 13915.55 ns/s


 url https://github.com/eric-wieser/cocotb/
 commit: dict-optimizations
 path: /tmp/cocotb-eric-wieser
 results: 
    simtime: 270162.40 ns
    realtime: 19.66s
    performance: 13744.99 ns/s
```
performance "improvement": -1.5%

**Comment by imphil** [comment]
ok, neither 2.7 nor 3.5 should see any improvements, so we have a bit of measurement uncertainty going on here. But it still looks like a rather large win on 3.7, nice!

**Comment by imphil** [comment]
@eric-wieser Thanks!

**Comment by imphil** [comment]
Thinking about it: is there a fast way to detect cpython, which has ordered `dict` since 3.6? (Reason for asking: 3.6 is the default in Ubuntu 18.04 LTS, hence widely use in the coming years.)

**Pull request merged**